TNL-3478 Cohort Membership race condition fix

efischer19 opened the pull request:
## [TNL-3478](https://openedx.atlassian.net/browse/TNL-3478)

Prevents users from joining more than one cohort group per course at the database level.

### Sandbox
- [x] http://efischer19.sandbox.edx.org/

### Testing
- [x] i18n (N/A, this is a backend change)
- [x] RTL (N/A, this is a backend change)
- [x] Unit, integration, acceptance tests as appropriate (will get thumbs on test plan before checking)
- [x] Analytics
- [x] Database migrations are backwards-compatible

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens 
- [x] Code review: @dianakhuang 
- [x] Product review: @explorerleslie 
- [x] Devops review: @maxrothman 

### Devops assistance
I've been talking to Max, he's going to review/support the migration plan in particular.

FYI: @ormsbee, Andy said you may have some input to add here regarding CourseUserGroup. 

### Post-review
- [x] Remove debugging code (`print`s and `time.sleep()`s)
- [x] Squash commits

efischer19 commented:
# The big picture
Users are ending up in multiple cohorts in a given course, which should not be happening.

## Some background
This situation was noted and fixed previously in https://openedx.atlassian.net/browse/TNL-1331, git changes: https://github.com/edx/edx-platform/commit/969bfa59166edcfe551a4d8e1348c3607e4fef56. However, this did not fix all instances of the problem, as noted in the new ticket. We may make use of the `remove_users_from_multiple_cohorts` command in fixing this.

## Where things stand today
There are a number (181, as of 16 Oct) of users in a bad state of belonging to more than one cohort. The extent of impact here is unknown, but this is a fairly core assumption being violated, so we can assume it's bad.

## How did this happen?
In local devstack testing, I was able to add some `time.sleep()` statements and trigger a race condition using multiple processes. Because the "update a user's cohort" operation was not implemented atomically, it was possible for a situation like this to happen:
* `Process 1`: Hey what cohort is user `FooBar` a member of?
* `Database to Process 1`: User `FooBar` is not in a cohort group.
* `Process 2`:  Hey what cohort is user `FooBar` a member of?
* `Database to Process 1`: User `FooBar` is not in a cohort group.
* `Process 1`: Okay, since there's no existing membership, go ahead and add user `Foobar` to cohort group `I Love Cats`
* `Database`: Recorded.
* `Process 2`: Okay, since there's no existing membership, go ahead and add user `Foobar` to cohort group `I Love Dogs`
* `Database`: Recorded.
At this point, according to the database, the user belongs to 2 cohorts, despite both process individually verifying that was not the case.

## The Fix
What I've elected to do as a fix is to create a new model class, `CohortMembership`. This class has a uniqueness constraint, enforced at the database level, that will never allow a user to have more than 1 CohortMembership in a given course. In addition, changes to this membership are made using a `filter(...).update(...)` call that increments the membership's version, so as to prevent more than 1 simultaneous process updating the value simultaneously.

I'll expand on the approach in more detail inline as needed.

## Tests
I wrote a test to confirm the pre-existing bad behavior, it's included in this PR. It does not work as expected though. Due to a bug/feature of Django, sharing an in-memory database across threads does not work (https://code.djangoproject.com/ticket/12118). This will be fixed on upgrading to Django 1.8. Until that point, the threads crash and the exception is never triggered, so the test passes at least. I'm open to not including the test in this PR, and adding it once the django upgrade is complete.

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -140,7 +145,6 @@ def get_cohorted_commentables(course_key):
     return ans
 
 
-@transaction.commit_on_success
```
All the transaction handling is now done inside `membership.save()`, it' no longer needed here.

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -202,7 +206,8 @@ def get_cohort(user, course_key, assign=True, use_cached=False):
 
-    user.course_groups.add(cohort)
+    membership = CohortMembership(course_user_group=cohort, user=user)
+    membership.save()
```
This is the way all cohort memberships should be added. Previously, it varied as to whether the `user` would add a new entry to its `course_groups`, or the `course_group` would add a new entry to its `user`. Only one way to do it now.

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -253,7 +258,9 @@ def get_course_cohorts(course, assignment_type=None):
-    query_set = CourseUserGroup.objects.filter(
+    query_set = CourseUserGroup.objects.exclude(
+        name=CourseUserGroup.INTERNAL_NAME
+    ).filter(
```
I'll expand on `CourseUserGroup.INTERNAL_NAME` below, these lines are to exclude it from appearing in the list of user-visible cohorts.

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -343,25 +350,19 @@ def add_user_to_cohort(cohort, username_or_email):
-            previous_cohort = course_cohorts[0]
-            previous_cohort.users.remove(user)
-            previous_cohort_name = previous_cohort.name
-            previous_cohort_id = previous_cohort.id
```
Both the `ValueError` and `previous_cohort` handling are now internal to `save()`

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -343,25 +350,19 @@ def add_user_to_cohort(cohort, username_or_email):
+        -remove previous info from "requested" event
+        -wait until after save so previous information is available
+        -add another event type and emit both before and after
+    """
```
Open TODO. Could affect analytics.

efischer19 commented on openedx/core/djangoapps/course_groups/migrations/0005_cohort_membership.py:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    complete_apps = ['course_groups']
```
This file was autogenerated to add the `CohortMembership` table.

efischer19 commented on openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py:
```diff
@@ -0,0 +1,94 @@
+class Migration(SchemaMigration):
+
+    def forwards(self, orm):
+        pass
```
Note this is currently an empty placeholder, I'll be handcrafting some SQL before we merge this PR into master.

efischer19 commented:
Logging a successful test run here for posterity. I first upload a csv file placing a user in one group, then click "add user" to synchronously place them in a different group. Note that this discrepancy is caught, and instead retried to keep the user's membership always in a good state.

```
2015-10-21 16:46:20,234 WARNING 29092 [instructor_task.api_helper] api_helper.py:79 - No duplicate tasks found: task_type cohort_students, task_key , and most recent task_id = 21
2015-10-21 16:46:20,275 INFO 29092 [edx.celery.task] tasks_helper.py:273 - Task: fd75a821-2ae4-4e52-905b-5a3bde3debff, InstructorTask ID: 22, Course: course-v1:edX+DemoX+Demo_Course, Input: {u'file_name': u'edX_DemoX_Demo_Course_cohorts_2015-10-21-204620.csv'}, Starting update (nothing cohorted yet)
Trying to add user to test 1
entering loop, trying to get membership for user 3, course course-v1:edX+DemoX+Demo_Course
No saved_membership!
Trying to add user to test 2
entering loop, trying to get membership for user 3, course course-v1:edX+DemoX+Demo_Course
No saved_membership!
User verified now in group test 1 in course course-v1:edX+DemoX+Demo_Course. Now at version
2015-10-21 16:46:30,329 INFO 29092 [edx.celery.task] tasks_helper.py:292 - Task: fd75a821-2ae4-4e52-905b-5a3bde3debff, InstructorTask ID: 22, Course: course-v1:edX+DemoX+Demo_Course, Input: {u'file_name': u'edX_DemoX_Demo_Course_cohorts_2015-10-21-204620.csv'}, Task type: cohorted, Finishing task: {'skipped': 0, 'succeeded': 1, 'attempted': 1, 'action_name': 'cohorted', 'failed': 0, 'step': 'Uploading CSV', 'duration_ms': 10052, 'total': 1}
[21/Oct/2015 16:46:30] "POST /courses/course-v1:edX+DemoX+Demo_Course/instructor/api/add_users_to_cohorts HTTP/1.1" 204 0
could not save membership, trying again
entering loop, trying to get membership for user 3, course course-v1:edX+DemoX+Demo_Course
User verified now in group test 2 in course course-v1:edX+DemoX+Demo_Course. Now at version
[21/Oct/2015 16:46:41] "POST /courses/course-v1:edX+DemoX+Demo_Course/cohorts/2/add HTTP/1.1" 200 231
[21/Oct/2015 16:46:41] "GET /courses/course-v1:edX+DemoX+Demo_Course/cohorts/ HTTP/1.1" 200 353
```

efischer19 commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,111 @@ def create(cls, name, course_id, group_type=COHORT):
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        self.full_clean(validate_unique=False)
```
The reasoning for not calling `validate_unique` is that we do allow multiple cohorts per user/course combination in memory, but not at the database level. The save method will call an update on any preexisting entries, so we'll still only have one saved to database at any given time.

cahrens replied:
Why do we want to allow multiple cohorts in memory?

efischer19 replied:
If we don't, then an error gets thrown because we have the new value (that we're trying to save) and the old value (currently in the database) in memory at the same time. However, I needed to have the other validation code in place to ensure `CohortMembership` is only ever used with `CourseUserGroup` objects of the appropriate type.

efischer19 commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,111 @@ def create(cls, name, course_id, group_type=COHORT):
+        print "Trying to add user to {}".format(self.course_user_group.name)
+
+        trying_to_save = True
+        while trying_to_save:
```
This loop is what makes simultaneous save attempts all succeed, rather than having one succeed and the rest fail.

efischer19 commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,111 @@ def create(cls, name, course_id, group_type=COHORT):
+                self.previous_cohort_name = saved_membership.course_user_group.name
+                self.previous_cohort_id = saved_membership.course_user_group.id
+
+            time.sleep(10)
```
This is necessary for being able to reproduce multiprocess errors locally.

cahrens replied:
Are you saying it is necessary for the unit test? I'm not clear why it is in the PR.

doctoryes replied:
If this sleep is necessary for testing only, it should be injected somehow during the test itself (preferably with mock).

efischer19 replied:
I left this in the PR solely for reviewers to be able to test locally if they so desired, I had intended to remove it before merging.

That's a good point though @doctoryes, I'll try to get this injected into tests to ensure those hit the race condition and handle it properly.

efischer19 commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,111 @@ def create(cls, name, course_id, group_type=COHORT):
+                        ).update(
+                            course_user_group=self.course_user_group,
+                            version=saved_membership.version + 1
+                        )
```
this `filter().update()` call is the heart of our new multiprocess safety. This will get transformed into one SQL call under-the-hood, and since it increments the `version` field, no two simultaneous queries will ever be able to match the same id/version combination.

doctoryes replied:
Right - what you really need here is an UPSERT, which Django ORM doesn't support without resorting to raw SQL. This solution is clever and might work. Do you know the exact SQL that gets generated by this compound ORM statement? Also, do you know if the same SQL is generated in tests (sqlite) and in stage/prod (MySQL)?

efischer19 replied:
I'm not sure what the exact SQL being generated is, in any environment. Is there any way to capture/inspect that while running?

I got the approach from this SO post, which seems to indicate it will be transformed into a single UPDATE: http://stackoverflow.com/questions/320096/django-how-can-i-protect-against-concurrent-modification-of-database-entries

efischer19 replied:
@doctoryes, here's some logging I was able to get locally, it's generating an `UPDATE ... WHERE` statement

https://github.com/edx/edx-platform/pull/10317

efischer19 commented:
This is ready for review, @cahrens and @dianakhuang would you mind reviewing?

FYI - @ormsbee, Andy told me that you may also be interested or have some input on this, having worked with CourseUserGroups in the past.

Heads up alert: @maxrothman, I haven't yet written the data migration, but this is the new PR where I will be doing so. I've changed my approach a bit from how it looked the other day, things are a bit simpler. I plan to finish writing the migration tomorrow, I'll ping you once it's done.

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -343,25 +348,19 @@ def add_user_to_cohort(cohort, username_or_email):
+        -remove previous info from "requested" event
+        -wait until after save so previous information is available
+        -add another event type and emit both before and after
+    """  # pylint: disable=W0105
```
@explorerleslie, I'm not sure if this is more of a product or an Analytics question.

Before, we got the `previous_cohort` information, emitted the `add_requested` event, then tried to save. Now, getting that information is a part of `save()`, so we'll have to choose from one of the options above.

nedbat replied:
Wouldn't it be easier to just use real comment syntax instead of adding a pylint pragma?

cahrens commented:
@doctoryes This is the PR I was talking about that enforces database transaction/locking. Can someone from Platform take a look at it?

cahrens commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -343,25 +348,19 @@ def add_user_to_cohort(cohort, username_or_email):
+
+    Decide how to best handle this
+        -remove previous info from "requested" event
+        -wait until after save so previous information is available
```
Am I correct that the only loss in functionality is that we won't emit a user_add_requested event if the save operation failed? What are the conditions under which it would fail?

efischer19 replied:
My main concern arose after I wrote these inline comments and found this code (https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/course_groups/cohorts.py#L38), which seems to be emitting an "action happened" event, in contrast to "the "action requested" event here. Could this timing information be getting gathered and utilized elsewhere? If that's the case, our best course of action would be to move the `previous_cohort` information to that event, and continue to emit the "requested" event prior to attempting the save.

Who would be the best person to ask about this design? @explorerleslie? someone from Analytics?

cahrens commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -202,7 +206,8 @@ def get_cohort(user, course_key, assign=True, use_cached=False):
         ).course_user_group
 
-    user.course_groups.add(cohort)
+    membership = CohortMembership(course_user_group=cohort, user=user)
```
With CohortMembership existing, should user.course_groups change (be deleted or changed to redirect to CohortMembership?). Same question about cohort.users.

efischer19 replied:
Ideally, I'd like to have `CourseUserGroup` and `User` relate to each other using a `ManyToManyField` with a `through` relationship to use `CohortMembership` as an intermediary. The problem there is that `CohortMembership` only applies to a subset of `CourseUserGroup` objects (those with `group_type=CourseUserGroup.COHORT`), and I don't think we can "filter" a `through` relationship. If we were to link them in this way, we'd end up assuming all `CourseUserGroup` objects were cohorts, which could break things if we ever end up making a different type that may not have the same uniqueness restrictions as `CourseUserGroup.COHORT`.

cahrens commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,112 @@ def create(cls, name, course_id, group_type=COHORT):
+        #print "Trying to add user to {}".format(self.course_user_group.name)
+
+        trying_to_save = True
+        while trying_to_save:
```
This while loop scares me. Why do we want to do this? But about the risk that we never exit?

efischer19 replied:
This is how I allow a retry of the save operation in case of race condition issues, as opposed to just failing. You're right about the risk it never exits though, I'll add a counter to allow it to fail out after a reasonable number of attempts.

doctoryes commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,112 @@ def create(cls, name, course_id, group_type=COHORT):
+            time.sleep(10)
+
+            try:
+                with transaction.commit_on_success():
```
`commit_on_success` has been removed in Django 1.8. Of course, you're not developing in 1.8 yet! We'll need to discuss how to move this over to `transaction.atomic` for the upgrade. We've already performed this for all other `commit_on_success

efischer19 replied:
There are 2 other points in this PR that explicitly need changes to work on django 1.8.

https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/models.py#L70
and
https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/models.py#L104

They're both fairly simple changes that I tried to outline with inline comments. I'd be glad to help get this code up-to-date when the time comes.

efischer19 replied:
Oh! Also, there are some issues worth metnioning with the test on this PR and django 1.8, outlined in the commit message https://github.com/edx/edx-platform/commit/5a4c0b3a8111fa3e31fef20d9464ff4417f46aad

nedbat commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +61,112 @@ def create(cls, name, course_id, group_type=COHORT):
+    class Meta(object):
+        unique_together = (('user', 'course_id'), )
+
+    #the sole purpose of overriding this method is to get the django 1.6 behavior of allowing 'validate_unique'
```
Sorry to be picky, but would you mind adding a space after these hash marks, and capitalizing the first word of sentences, so we can more easily read the comment?

efischer19 commented on openedx/core/djangoapps/course_groups/cohorts.py:
```diff
@@ -369,12 +358,11 @@ def add_user_to_cohort(cohort, username_or_email):
+            "previous_cohort_id": membership.previous_cohort_id,
+            "previous_cohort_name": membership.previous_cohort_name,
         }
     )
```
@explorerleslie, is this a decision you can help with? The behavior of this event has changed slightly; it is no longer possible to emit `user_add_requested` with previous cohort information before the membership is saved. We'll have to either remove the previous cohort information, or emit the event after the membership is saved.

The previous lines containing a TODO have been removed, here's the conversation that had happened to that point:
```
TODO: with the new setup, we cannot emit a "requested" event containing
information about the previous cohort before calling save().

Decide how to best handle this
    -remove previous info from "requested" event
    -wait until after save so previous information is available
    -add another event type and emit both before and after
```

cahrens   Collaborator
cahrens added a note a day ago
Am I correct that the only loss in functionality is that we won't emit a user_add_requested event if the save operation failed? What are the conditions under which it would fail?

 efischer19   Collaborator
efischer19 added a note a day ago
My main concern arose after I wrote these inline comments and found this code (https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/course_groups/cohorts.py#L38), which seems to be emitting an "action happened" event, in contrast to "the "action requested" event here. Could this timing information be getting gathered and utilized elsewhere? If that's the case, our best course of action would be to move the previous_cohort information to that event, and continue to emit the "requested" event prior to attempting the save.

Who would be the best person to ask about this design? explorerleslie? someone from Analytics?

efischer19 replied:
Spoke offline with @explorerleslie, we'll be emitting the same event, with previous cohort information, after the `save()` call returns.

efischer19 commented:
Reposting force-push-obliterated comment from an earlier commit to keep it from being lost, relevant link: https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py

@maxrothman, here's what I've come up with for the data migration. Per @explorerleslie, we can arbitrarily assign "bad state" users to either of their current memberships.

I've also elected to make the backwards migration a no-op like we discussed earlier. I'm still not entirely sure if there's more we need to do re: your questions about migrating from scratch, won't `for cohort_group in orm.CourseUserGroup.objects.all():` above effectively turn into an iteration over `range(0,0)` and do nothing?

I'm also curious if you'd have any thoughts/recommendations when we actually perform these migrations - should we do this entire thing transactionally with a table lock, to prevent more bad situations being created while we're fixing the current ones? My current thought is that if we apply the 0005 migration, then push the code changes, we'll have protection against double-enrollments, and at that point we can run the 0006 migration to fix previously-created errors.

efischer19 commented:
jenkins run all

efischer19 commented:
@cahrens @dianakhuang @doctoryes, this PR is ready for re-review when you get a chance. I've addressed the comments made previously and changes some pre-existing unit tests to get them working with this new `CohortMembership` paradigm.

I've also removed the clunky `sleep()` lines and test from this PR, and have created a separate (under development) PR that will allow for much more reliable unit tests of the race condition handling here: https://github.com/edx/edx-platform/pull/10340

efischer19 commented:
jenkins run js

maxrothman commented:
@efischer19 re the data migration: is it not in this PR?

maxrothman commented on openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py:
```diff
@@ -0,0 +1,116 @@
+                unioned_set = set(current_course_groups).union(set(current_user_groups))
+
+                # Per product guidance, fix problem users by arbitrarily choosing a single membership to retain
+                cohort = unioned_set[0]
```
Have you tested this? Sets don't support indexing.

efischer19 replied:
I have not yet tested this, but I will this afternoon. `pop()` should do what I want here, correct?

maxrothman replied:
Is the goal to pick one arbitrarily? If so, yes.

efischer19 replied:
Great, thanks

maxrothman commented on openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py:
```diff
@@ -0,0 +1,116 @@
+                # Per product guidance, fix problem users by arbitrarily choosing a single membership to retain
+                cohort = unioned_set[0]
+
+                for i in range(1, len(unioned_set)):
```
Nitpick: `for i, thing in enumerate(unioned_set):` would be more idiomatic.

doctoryes commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +59,96 @@ def create(cls, name, course_id, group_type=COHORT):
+                    self.user.course_groups.add(self.course_user_group)
+                    if saved_membership is not None:
+                        self.previous_cohort.users.remove(self.user)
+                        updated = CohortMembership.objects.filter(
```
Why not use a `select_for_update()` here?:

https://docs.djangoproject.com/en/1.4/ref/models/querysets/#select-for-update

If you're certain that a single `UPDATE..WHERE` SQL statement is generated by the `filter().update()`, then it seems this code will work. But `select_for_update()` seems designed for the situation where the code wants to read a row/rows and update them within a transaction. (Just wondering...)

dianakhuang commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +59,96 @@ def create(cls, name, course_id, group_type=COHORT):
+
+    previous_cohort = None
+    previous_cohort_name = None
+    previous_cohort_id = None
```
I believe by convention, these fields are placed at the top of the class, just so it's easier to understand what these values are when they're used.

dianakhuang commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +59,96 @@ def create(cls, name, course_id, group_type=COHORT):
+        self.full_clean(validate_unique=False)
+
+        max_retries = 5
+        for num_attempts in range(max_retries):
```
I would also add a comment explaining why we're allowing for 5 attempts.

efischer19 commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +60,96 @@ def create(cls, name, course_id, group_type=COHORT):
+                    saved_membership = CohortMembership.objects.select_for_update().get(
+                        user__id=self.user.id,  # pylint: disable=E1101
+                        course_id=self.course_id
+                    )
```
@doctoryes, how's this new approach look? It seems to handle race conditions correctly (tested using the `sleep()` calls you see here), and it looks simpler, to me at least.

efischer19 commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +60,96 @@ def create(cls, name, course_id, group_type=COHORT):
+
+                    saved_membership.course_user_group=self.course_user_group
+
+                    #note: in django 1.8, we can call save with updated_fields=['course_user_group']
```
should say `update_fields`, my bad on the typo

efischer19 commented:
@cahrens @dianakhuang - I've addressed your previous comments, mind re-reviewing when you get a chance?

Spoke with @doctoryes offline about `get_or_create()`, everything looks alright there.

@maxrothman - new data migration tests are forthcoming, I'm trying to detangle a mess of migrations locally right now.

Commits have been squashed, what you see is what I plan to merge, pending the test PR.

efischer19 commented:
jenkins run lettuce

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run bokchoy

cahrens commented on openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py:
```diff
@@ -0,0 +1,115 @@
+        for cohort_group in orm.CourseUserGroup.objects.all():
+            for user in cohort_group.users.all():
+                current_course_groups = orm.CourseUserGroup.objects.filter(
+                    course_id=cohort_group.course_id,
```
Can be in multiple CourseUserGroups if not cohort...

efischer19 commented on requirements/edx/base.txt:
```diff
@@ -120,6 +120,7 @@ django_debug_toolbar==1.3.2
 
 # Used for testing
 astroid==1.3.8
+before_after==0.1.3
```
See https://github.com/edx/edx-platform/pull/10340 for details of this library being added.

efischer19 commented:
@maxrothman I've updated the gist at https://gist.github.com/efischer19/e7d892654210ec04a458 with details of the fixed-up migration. Changes made were to get `course_id` migrated properly (as we noted earlier), and to filter the migrated data to only `CourseUserGroup`s of type `cohort`, as mentioned by @cahrens 

All reviewers - I merged in the thumbs'd and test-passing PR for the before_after unit test and there are no further changes planned, so this is now the exact diff I'll be merging into master.

efischer19 commented:
jenkins run python

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
jenkins run lettuce

cahrens commented on openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py:
```diff
@@ -0,0 +1,117 @@
+                    users__id=user.id,
+                    group_type=cohort_type
+                )
+                current_user_groups = user.course_groups.filter(
```
Do you also need to add filtering here for cohort_type?

efischer19 replied:
Yeah, good catch

cahrens commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +59,99 @@ def create(cls, name, course_id, group_type=COHORT):
+        max_retries = 2
+        success = False
+        for dummy_counter in range(max_retries):
+            saved_membership = None
```
Nit: is there a reason why you set saved_membership to None originally? You don't do this for created, and I'm having trouble understanding the difference.

efischer19 replied:
No, it's a holdover from before I added `get_or_create()`, when I had to test for `None` to see if anything was returned. I'll remove it.

cahrens commented:
:+1: after you add filtering (if necessary)

doctoryes commented on openedx/core/djangoapps/course_groups/models.py:
```diff
@@ -58,6 +59,99 @@ def create(cls, name, course_id, group_type=COHORT):
+        # The limit is 2, since select_for_update ensures atomic updates. Creation is the only possible race condition.
+        max_retries = 2
+        success = False
+        for dummy_counter in range(max_retries):
```
Nit: The standard I've seen is to replace `dummy_counter` with a `__`.

dianakhuang commented:
:+1: from me as well.

maxrothman commented on openedx/core/djangoapps/course_groups/migrations/0006_cohort_membership_data_migrate.py:
```diff
@@ -0,0 +1,118 @@
+                unioned_set = set(current_course_groups).union(set(current_user_groups))
+
+                # Per product guidance, fix problem users by arbitrarily choosing a single membership to retain
+                cohort_itr = unioned_set.pop()
```
Nitpick: using the same name for this and the iterator variable below is slightly confusing.

efischer19 replied:
I'm glad you spotted this - shouldn't we be saving this value? When we create the actual membership a few lines below, we should be using this value instead of the outer loop iterator `cohort_group`.

maxrothman commented:
:+1:

efischer19 commented:
@maxrothman I've remembered why I wanted to roll this out as 0005 migration -> code change -> 0006 migration. If we run both migrations and keep running old code against that, then changes made to cohort memberships will not be properly tracked in the `CohortMembership` table, because there's no code enforcing that relationship.

However, rolling out the 0005 migration first will provide the data needed for new code to run, which will then be tracking any changes made to membership in the new system. At that point we can safely roll out 0006 and have everyone in a good state.

Gist showing all of this: https://gist.github.com/efischer19/d62f8ee42b7fbfbc6c9a

Is a partial migration rollout like that something that's possible for us to do? Am I worrying too much about changes made in a small window?

efischer19 commented on openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh:
```diff
@@ -0,0 +1,8 @@
+fi
+
+./manage.py lms migrate --settings=$1 course_groups 0005 --fake
+./manage.py lms migrate --settings=$1 course_groups 0006
```
@maxrothman, how's this look?

Sample output:
```
edxapp@precise64:~/edx-platform$ sh openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh devstack
 - Soft matched migration 0005 to 0005_cohort_membership.
Running migrations for course_groups:
 - Migrating backwards to just after 0005_cohort_membership.
 < course_groups:0006_cohort_membership_data_migrate
   (faked)
 - Soft matched migration 0006 to 0006_cohort_membership_data_migrate.
Running migrations for course_groups:
 - Migrating forwards to 0006_cohort_membership_data_migrate.
 > course_groups:0006_cohort_membership_data_migrate
 - Migration 'course_groups:0006_cohort_membership_data_migrate' is marked for no-dry-run.
 - Loading initial data for course_groups.
Installed 0 object(s) from 0 fixture(s)
edxapp@precise64:~/edx-platform$ sh openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh
openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh: usage: rerun_0006.sh environment
```

maxrothman replied:
Users might want to append other options too (e.g. `db-dry-run`), so why not just throw all the arguments on the end?

efischer19 commented on openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh:
```diff
@@ -0,0 +1,8 @@
+fi
+
+./manage.py lms migrate course_groups 0005 --fake "$@"
+./manage.py lms migrate course_groups 0006 "$@"
```
@maxrothman Oh yeah, great point. This look better?

```
edxapp@precise64:~/edx-platform$ sh openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh
openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh: usage: rerun_0006.sh <arguments>. At minimum, '--settings=<environment>' is expected.
edxapp@precise64:~/edx-platform$ sh openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh --settings=devstack
 - Soft matched migration 0005 to 0005_cohort_membership.
Running migrations for course_groups:
 - Migrating backwards to just after 0005_cohort_membership.
 < course_groups:0006_cohort_membership_data_migrate
   (faked)
 - Soft matched migration 0006 to 0006_cohort_membership_data_migrate.
Running migrations for course_groups:
 - Migrating forwards to 0006_cohort_membership_data_migrate.
 > course_groups:0006_cohort_membership_data_migrate
 - Migration 'course_groups:0006_cohort_membership_data_migrate' is marked for no-dry-run.
 - Loading initial data for course_groups.
Installed 0 object(s) from 0 fixture(s)
```

maxrothman replied:
Don't think you want quotes on the $@ there.

On Fri, Oct 30, 2015 at 11:09 AM, Eric Fischer <notifications@github.com>
wrote:

> In openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh
> <https://github.com/edx/edx-platform/pull/10295#discussion_r43512369>:
>
> > @@ -0,0 +1,8 @@
> > +#!/bin/bash
> > +if [ $# -eq 0 ]; then
> > +    echo "$0: usage: rerun_0006.sh <arguments>. At minimum, '--settings=<environment>' is expected."
> > +    exit 1
> > +fi
> > +
> > +./manage.py lms migrate course_groups 0005 --fake "$@"
> > +./manage.py lms migrate course_groups 0006 "$@"
>
> @maxrothman <https://github.com/maxrothman> Oh yeah, great point. This
> look better?
>
> edxapp@precise64:~/edx-platform$ sh openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh
> openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh: usage: rerun_0006.sh <arguments>. At minimum, '--settings=<environment>' is expected.
> edxapp@precise64:~/edx-platform$ sh openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh --settings=devstack
>  - Soft matched migration 0005 to 0005_cohort_membership.
> Running migrations for course_groups:
>  - Migrating backwards to just after 0005_cohort_membership.
>  < course_groups:0006_cohort_membership_data_migrate
>    (faked)
>  - Soft matched migration 0006 to 0006_cohort_membership_data_migrate.
> Running migrations for course_groups:
>  - Migrating forwards to 0006_cohort_membership_data_migrate.
>  > course_groups:0006_cohort_membership_data_migrate
>  - Migration 'course_groups:0006_cohort_membership_data_migrate' is marked for no-dry-run.
>  - Loading initial data for course_groups.
> Installed 0 object(s) from 0 fixture(s)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/10295/files#r43512369>.
>

efischer19 replied:
Why is that? (I'm not super well-versed in bash scripting in general).

Each positional argument should be split into its own separate word: http://www.gnu.org/software/bash/manual/bash.html#Special-Parameters

maxrothman replied:
Never mind, you're correct.

On Fri, Oct 30, 2015 at 11:15 AM, Eric Fischer <notifications@github.com>
wrote:

> In openedx/core/djangoapps/course_groups/migrations/rerun_0006.sh
> <https://github.com/edx/edx-platform/pull/10295#discussion_r43513157>:
>
> > @@ -0,0 +1,8 @@
> > +#!/bin/bash
> > +if [ $# -eq 0 ]; then
> > +    echo "$0: usage: rerun_0006.sh <arguments>. At minimum, '--settings=<environment>' is expected."
> > +    exit 1
> > +fi
> > +
> > +./manage.py lms migrate course_groups 0005 --fake "$@"
> > +./manage.py lms migrate course_groups 0006 "$@"
>
> Why is that? (I'm not super well-versed in bash scripting in general).
>
> Each positional argument should be split into its own separate word:
> http://www.gnu.org/software/bash/manual/bash.html#Special-Parameters
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/10295/files#r43513157>.
>

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
jenkins run lettuce

efischer19 commented:
jenkins run python

efischer19 commented:
The tests are being super goofy this morning. They've previously passed for all checks, and the only change I've made since then is the small bash script I added in `migrations`. Merging.
